torch/csrc/distributed/c10d/CudaDMAConnectivity.cpp

```diff
@@ -0,0 +1,118 @@
+#if !defined(USE_ROCM) && defined(PYTORCH_C10_DRIVER_API_SUPPORTED)
+#include <torch/csrc/distributed/c10d/DMAConnectivity.hpp>
+
+#include <c10/cuda/CUDAException.h>
+#include <c10/cuda/driver_api.h>
+
+#include <cuda_runtime.h>
+#include <nvml.h>
+
+namespace {
+
+std::string get_bus_id(int device_idx) {
+  char bus_id[80];
+  cudaDeviceProp prop{};
+  C10_CUDA_CHECK(cudaGetDeviceProperties(&prop, device_idx));
+  snprintf(
+      bus_id,
+      sizeof(bus_id),
+      NVML_DEVICE_PCI_BUS_ID_FMT,
+      prop.pciDomainID,
+      prop.pciBusID,
+      prop.pciDeviceID);
+  return std::string(bus_id);
+}
+
+struct C10_EXPORT NVLinkDetector : public c10d::DMAConnectivityDetector {
+  c10::intrusive_ptr<c10d::DMAConnectivity> detect() override {
+    int num_devices;
+    C10_CUDA_CHECK(cudaGetDeviceCount(&num_devices));
+
+    std::vector<std::vector<int>> matrix;
+    matrix.reserve(num_devices);
+    for (int i = 0; i < num_devices; ++i) {
+      matrix.emplace_back(num_devices, 0);
+    }
+
+    // Obtain the bus_id for all visible devices
+    std::unordered_map<std::string, int> bus_id_to_device_idx;
+    std::vector<std::string> bus_ids;
+    bus_ids.reserve(num_devices);
+    for (int i = 0; i < num_devices; ++i) {
+      auto bus_id = get_bus_id(i);
+      bus_id_to_device_idx.emplace(bus_id, i);
+      bus_ids.push_back(std::move(bus_id));
+    }
+
+    // Obtain the nvml device for all bus_ids
+    auto driver_api = c10::cuda::DriverAPI::get();
+    std::vector<nvmlDevice_t> nvml_devices(num_devices, nullptr);
+    for (int i = 0; i < num_devices; ++i) {
+      TORCH_CHECK_EQ(
+          driver_api->nvmlDeviceGetHandleByPciBusId_v2_(
+              bus_ids[i].c_str(), &nvml_devices[i]),
+          NVML_SUCCESS);
+    }
+
+    std::vector<int> switch_link_count(num_devices, 0);
+    for (int i = 0; i < num_devices; ++i) {
+      for (int link = 0; link < 20; ++link) {
```

Comment:
Is `20` arbitrary with no meaning or something meaningful?

Reply:
Good question. As far as I know, there's no nvml API that allows us to query the number of NVLinks. So we just increment the link id until it fails. Here we just want a reasonably large number.

I'll change this to a constant and add a comment.